presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -614,4 +615,66 @@ private static boolean isScheduled(Optional<StageInfo> rootStage)
                 .map(StageInfo::getState)
                 .allMatch(state -> (state == StageState.RUNNING) || state.isDone());
     }
+
+    public Optional<QueryInfo> getFinalQueryInfo()
+    {
+        if (finalQueryInfo.get() == null) {
+            return Optional.empty();
+        }
+        return Optional.of(finalQueryInfo.get());
+    }
+
+    public QueryInfo updateQueryInfo(Optional<StageInfo> stageInfo)
+    {
+        QueryInfo queryInfo = getQueryInfo(stageInfo);
+        if (queryInfo.isFinalQueryInfo()) {
+            finalQueryInfo.compareAndSet(null, queryInfo);
+        }
+        return queryInfo;
+    }
+
+    public void pruneQueryInfo()
+    {
+        if (finalQueryInfo.get() == null || !finalQueryInfo.get().getOutputStage().isPresent()) {
+            return;
+        }
+        QueryInfo queryInfo = finalQueryInfo.get();
```

Comment:
I'd move this line above the `if` and use the value in the condition, so it only needs to be read once